libsawtooth/src/client/rest.rs

```diff
@@ -148,40 +159,107 @@ impl Iterator for BatchIter {
     }
 }
 
+/// Iterator used for parsing and deserializing transactions.
+struct TransactionIter {
+    next: Option<String>,
+    cache: VecDeque<Transaction>,
+}
+
+impl TransactionIter {
+    /// Create a new 'TransactionIter' which will make a call to the REST API and load the initial
+    /// cache with the first page of transactions.
+    fn new(url: &str) -> Result<Self, SawtoothClientError> {
+        let mut new_iter = Self {
+            next: Some(url.to_string()),
+            cache: VecDeque::with_capacity(0),
+        };
+        new_iter.reload_cache()?;
+        Ok(new_iter)
+    }
+
+    /// If another page of transactions exisits, use the 'next' url from the current page and
+    /// reload the cache with the next page of transactions.
+    fn reload_cache(&mut self) -> Result<(), SawtoothClientError> {
+        if let Some(url) = &self.next.take() {
+            let request = Client::new().get(url);
+            let response = request.send().map_err(|err| {
+                SawtoothClientError::new_with_source("request failed", err.into())
+            })?;
+
+            let page: TransactionPage = response.json().map_err(|err| {
+                SawtoothClientError::new_with_source(
+                    "failed to deserialize response body",
+                    err.into(),
+                )
+            })?;
+
+            self.cache = page.data.into();
+
+            self.next = page.next;
+        }
+        Ok(())
+    }
+}
+
+impl Iterator for TransactionIter {
+    type Item = Result<ClientTransaction, SawtoothClientError>;
+    /// Return the next transaction from the cache, if the cache is empty reload it.
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cache.is_empty() && self.next.is_some() {
+            if let Err(err) = self.reload_cache() {
+                return Some(Err(SawtoothClientError::new_with_source(
+                    "Unable to load iterator cache: {}",
+                    err.into(),
+                )));
+            }
```

Comment:
Probably best to avoid wrapping a `SawtoothClientError` in another `SawtoothClientError`. You could just `return Some(Err(err))` in the `if let`, or if you want to get fancy with it, you can do `self.reload_cache().map_err(Some).transpose()?;`.

Reply:
Changed to return Some(Err(err))